### ISSUE TITLE
list index out of range in freeze_graph.py

### ISSUE THREAD

**Issue by burui11087** [tensorflow/tensorflow#24591]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):**NO**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):**Linux Ubuntu 16.04.5**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:**No**
- TensorFlow installed from (source or binary):**binary**
- TensorFlow version (use command below):**1.12.0**
- Python version:**3.5.2**
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:**9.0/7.0**
- GPU model and memory:**GTX 1080 11GB**

**Describe the current behavior**
I try to freeze graph generated by tf.esitimator. esitimator will generate *.pbtxt and ckpt file automatically, so I use following command:
`python freeze_graph.py --input_graph=path/to/.pbtxt --input_checkpoint=path/to/model.ckpt-0 --output_graph=path/to/saved/.pb --output_node_name=OutPutOp`

**Other info / logs**
`Traceback (most recent call last):
  File "model_freeze.py", line 494, in <module>
    run_main()
  File "model_freeze.py", line 490, in run_main
    app.run(main=my_main, argv=[sys.argv[0]] + unparsed)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/platform/app.py", line 125, in run
    _sys.exit(main(argv))
  File "model_freeze.py", line 489, in <lambda>
    my_main = lambda unused_args: main(unused_args, flags)
  File "model_freeze.py", line 382, in main
    flags.saved_model_tags, checkpoint_version)
  File "model_freeze.py", line 364, in freeze_graph
    checkpoint_version=checkpoint_version)
  File "model_freeze.py", line 191, in freeze_graph_with_def_protos
    var_list=var_list, write_version=checkpoint_version)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 1102, in __init__
    self.build()
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 1114, in build
    self._build(self._filename, build_save=True, build_restore=True)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 1151, in _build
    build_save=build_save, build_restore=build_restore)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 773, in _build_internal
    saveables = self._ValidateAndSliceInputs(names_to_saveables)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 680, in _ValidateAndSliceInputs
    for converted_saveable_object in self.SaveableObjectsForOp(op, name):
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 654, in SaveableObjectsForOp
    variable, "", name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 128, in __init__
    self.handle_op = var.op.inputs[0]
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 2128, in __getitem__
    return self._inputs[i]
IndexError: list index out of range
`

**Comment by jvishnuvardhan** [comment]
@burui11087, 
Could you provide a code to reproduce the issue to find root-cause of the issue? 
Thank you

**Comment by jiayiliu** [comment]
@jvishnuvardhan Thank you for checking on this.  It links to couple other issues #22029 #5387 #4363

Here I provides a minimal code using tf hub:

```python
import tensorflow as tf
import tensorflow_hub as hub

module = hub.Module("https://tfhub.dev/google/imagenet/resnet_v2_50/classification/1")
saver = tf.train.Saver()
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    tf.train.write_graph(sess.graph, "resnet", "model.pb", as_text=False)
    saver.save(sess,'resnet/model')
```

Then I encountered the IndexError as I run

`python -m tensorflow.python.tools.freeze_graph --input_graph=resnet/model.pb --input_checkpoint=resnet/model --output_graph=resnet/frozen_graph.pb --output_node_names=module/resnet_v2_50/predictions/Reshape_1  --input_binary`

At the same time, i have successfully frozen a plain-vanilla single layer CNN.  Could you point some directions to check?
Thanks!

**Comment by bobye** [comment]
I have the same problem when using keras advanced layers.

**Comment by IrvingMeng** [comment]
same here.

**Comment by gauravlath07** [comment]
Encountered the same error while freezing transformer from tensorflow official models

**Comment by stevehawley** [comment]
I'm getting the same error on a keras implementation of yolo. 

The error occurs in `__init__` of a `ResourceVariableSaveable` when trying to access `var.op.inputs[0]` on a tensor:
```
Tensor("batch_normalization_1/beta:0", shape=(), dtype=resource)
```
op is a "VarHandleOp" and `var.op.inputs` is a `tensorflow.python.framework.ops.Operation._InputList` object.

The input list in this case is empty. Does the resource variable have to have an input? What should handle_op be in this case?

Any suggestions for how to proceed?

**Comment by jiayiliu** [comment]
@stevehawley  - have you tried with the tf.keras rather than the native keras?

**Comment by stevehawley** [comment]
@jiayiliu - yes, I am using the internal `tensorflow.keras`.

My workaround has been to use `tf.saved_model.simple_save` instead of a `Saver` and have freeze_graph use input_saved_model_dir. 

It gets me past this problem but I don't consider the issue resolved. If freeze_graph is not going to support working on normal graphs it should deprecate that api and update the docs, or at least put a note that for keras models you have to used saved_model.

**Comment by jiayiliu** [comment]
Thanks for sharing the workaround @stevehawley .  I agree with you that the problem is not solved, especially the `simple_save` is marked as DEPRECATED in the documentation.   I switched to keras.applications rather than using the tf.hub.

**Comment by andresusanopinto** [comment]
Unclear why this was assigned to me. TensorFlow Hub was used to produce a failing case, but IIUC the issue exists even without using hub.

**Comment by HariSeldon** [comment]
I am hitting this same problem too.
Is there any updates ? 
Is freeze_graph.py still officially supported ? 
Is there an alternative way of generating a protobuf out of a tf.graph ?

**Comment by nidetaoge** [comment]
I also meet this porblem when using mnasnet and efficientnet to freeze, where I download models from tensorflow/tpu models: https://github.com/tensorflow/tpu/tree/master/models/official. I use tensorflow 1.11.0 and I have checked my input/output node names which are right.

  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/tools/freeze_graph.py", line 363, in freeze_graph
    checkpoint_version=checkpoint_version)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/tools/freeze_graph.py", line 190, in freeze_graph_with_def_protos
    var_list=var_list, write_version=checkpoint_version)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 1094, in __init__
    self.build()
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 1106, in build
    self._build(self._filename, build_save=True, build_restore=True)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 1143, in _build
    build_save=build_save, build_restore=build_restore)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 765, in _build_internal
    saveables = self._ValidateAndSliceInputs(names_to_saveables)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 672, in _ValidateAndSliceInputs
    for converted_saveable_object in self.SaveableObjectsForOp(op, name):
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 646, in SaveableObjectsForOp
    variable, "", name)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/training/saver.py", line 128, in __init__
    self.handle_op = var.op.inputs[0]
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/ops.py", line 2126, in __getitem__
    return self._inputs[i]
IndexError: list index out of range

Can any one help me? Thx!

**Comment by VertexC** [comment]
I got same issue. Here is the [actual script](https://gist.github.com/thepulkitagarwal/36bc45aa43ae43f83baa5111a89be73e#file-freezekerasmodel-py.) help me to freeze the model. Not that you need to modify the output node name manually.  And I successfully loaded the frozen model onto android.

**Comment by stefan-falk** [comment]
I get the same issue when I try to freeze the bidirectional LSTM network from tensor2tensor (https://github.com/tensorflow/tensor2tensor/blob/v.1.12.0/tensor2tensor/models/lstm.py).

**Comment by G-mel** [comment]
I managed to freeze my tensorflow model with `tf.keras` layers in it by using `tf.SavedModel` in conjuntion with `freeze_graph.freeze_graph_with_def_protos`. Below is a generic snippet to convert a `tf.keras` based tensorflow checkpoint into a frozen checkpoint.

I am working in `tensorflow==1.13.1`
```
<create evaluation graph/model here>

export_dir = <SavedModel serving dir>
with <eval_graph>.as_default() as graph:
    saver = tf.train.Saver()
    init_op = tf.global_variables_initializer()
    
    with tf.Session(graph=graph) as sess:
        sess.run(init_op)
        saver.restore(sess, <checkpoint_path>)

        builder = tf.saved_model.Builder(export_dir)
        builder.add_meta_graph_and_variables(sess,
            [tf.saved_model.tag_constants.SERVING],
            strip_default_attrs=True)
    builder.save()

    from tensorflow.python.tools import freeze_graph
    freeze_graph.freeze_graph_with_def_protos(
        ...
        input_saved_model_dir=export_dir,
        saved_model_tags=[tf.saved_model.tag_constants.SERVING]
    )
```

**Comment by stoooops** [comment]
@G-mel What do you mean by `<eval_graph>` in your code? What type is that?

**Comment by sseveran** [comment]
@ymodak Should we expect any further updates to graph freezing? freeze_graph.py is still in master. Tensorflow has moved to Keras and the incompatibility should probably be either documented or fixed.

**Comment by ymodak** [comment]
Please test with the latest version of tensorflow for using `freeze_graph` tool. 
Thanks!

**Comment by johannabar** [comment]
I believe this is still not working with the latest version of tensorflow 2.1.0. The error becomes: `self.handle_op = var.op.inputs[0]
IndexError: tuple index out of range`
@ymodak Any updates on this?

The full stack error is:
```
 File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 491, in <module>
    run_main()
  File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 487, in run_main
    app.run(main=my_main, argv=[sys.argv[0]] + unparsed)
  File "/path/to/python3.6/site-packages/tensorflow_core/python/platform/app.py", line 40, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
  File "/path/to/python3.6/site-packages/absl/app.py", line 300, in run
    _run_main(main, args)
  File "/path/to/python3.6/site-packages/absl/app.py", line 251, in _run_main
    sys.exit(main(argv))
  File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 486, in <lambda>
    my_main = lambda unused_args: main(unused_args, flags)
  File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 378, in main
    flags.saved_model_tags, checkpoint_version)
  File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 361, in freeze_graph
    checkpoint_version=checkpoint_version)
  File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 190, in freeze_graph_with_def_protos
    var_list=var_list, write_version=checkpoint_version)
  File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saver.py", line 828, in __init__
    self.build()
  File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saver.py", line 840, in build
    self._build(self._filename, build_save=True, build_restore=True)
  File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saver.py", line 878, in _build
    build_restore=build_restore)
  File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saver.py", line 482, in _build_internal
    names_to_saveables)
  File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saving/saveable_object_util.py", line 349, in validate_and_slice_inputs
    for converted_saveable_object in saveable_objects_for_op(op, name):
  File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saving/saveable_object_util.py", line 210, in saveable_objects_for_op
    variable, "", name)
  File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saving/saveable_object_util.py", line 84, in __init__
    self.handle_op = var.op.inputs[0]
IndexError: tuple index out of range
```

**Comment by qipengh** [comment]
> I believe this is still not working with the latest version of tensorflow 2.1.0. The error becomes: `self.handle_op = var.op.inputs[0] IndexError: tuple index out of range`
> @ymodak Any updates on this?
> 
> The full stack error is:
> 
> ```
>  File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 491, in <module>
>     run_main()
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 487, in run_main
>     app.run(main=my_main, argv=[sys.argv[0]] + unparsed)
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/platform/app.py", line 40, in run
>     _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
>   File "/path/to/python3.6/site-packages/absl/app.py", line 300, in run
>     _run_main(main, args)
>   File "/path/to/python3.6/site-packages/absl/app.py", line 251, in _run_main
>     sys.exit(main(argv))
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 486, in <lambda>
>     my_main = lambda unused_args: main(unused_args, flags)
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 378, in main
>     flags.saved_model_tags, checkpoint_version)
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 361, in freeze_graph
>     checkpoint_version=checkpoint_version)
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/tools/freeze_graph.py", line 190, in freeze_graph_with_def_protos
>     var_list=var_list, write_version=checkpoint_version)
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saver.py", line 828, in __init__
>     self.build()
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saver.py", line 840, in build
>     self._build(self._filename, build_save=True, build_restore=True)
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saver.py", line 878, in _build
>     build_restore=build_restore)
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saver.py", line 482, in _build_internal
>     names_to_saveables)
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saving/saveable_object_util.py", line 349, in validate_and_slice_inputs
>     for converted_saveable_object in saveable_objects_for_op(op, name):
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saving/saveable_object_util.py", line 210, in saveable_objects_for_op
>     variable, "", name)
>   File "/path/to/python3.6/site-packages/tensorflow_core/python/training/saving/saveable_object_util.py", line 84, in __init__
>     self.handle_op = var.op.inputs[0]
> IndexError: tuple index out of range
> ```

Do you have any progress？I have the same problem.

**Comment by cmehrshad** [comment]
Same here. Tensorflow 2.1.0 and keras 2.3.1.

**Comment by qipengh** [comment]
@cmehrshad Maybe because of TF2.x use resource variables by default. My job is make export_inference_graph.py working in TF_models/slim/ with TF2.x. Then I try print var_list.value before the code line :`saver = saver_lib.Saver(var_list,write_version=checkpoint_version)`, the tensor like 'shape=(), dtype=resource' . when I use `tf.compat.v1.disable_resource_variables()`, error occurs  `NameError: gobal name 'distribute_strategy' is not defined`. so I haven't any solution now.

**Comment by weilanShi** [comment]
I guess the main reason is because of the use of keras，i change “tf.keras.layers.BatchNormalization ” to "tf.layers.batch_normalization"，and the freeze_graph is work.

**Comment by qshler** [comment]
Any updates on this? 
I have the exact same callstack as @qipengh.
Any updates on the issue? @ymodak

**Comment by 76176235** [comment]
Same here. Tensorflow 2.2.0

**Comment by ariqathallah38** [comment]
Is it fixed? Same here : Tensorflow 2.2

**Comment by zhenzhen2017** [comment]
**### _root cause:_**
maybe list index out of range in freeze_graph.py is caused by a bug in saveable_object_util.py

stack is:
https://github.com/tensorflow/tensorflow/blob/v1.14.0/tensorflow/python/tools/freeze_graph.py#L189
https://github.com/tensorflow/tensorflow/blob/v1.14.0/tensorflow/python/training/saver.py#L481
https://github.com/tensorflow/tensorflow/blob/v1.14.0/tensorflow/python/training/saving/saveable_object_util.py#L82 

![image](https://user-images.githubusercontent.com/33881637/127797539-b4032bab-d383-4176-80c3-48be6107e77e.png)

### **_solution:_**
use tf.graph_util.convert_variables_to_constants(sess, sess.graph.as_graph_def(add_shapes=True), FLAGS.output_names.split(","))   instead of freeze_graph.py.
